### PR TITLE
add slanext models

### PR DESCRIPTION
**Pull request by liu-jiaxuan**
新增最新表格识别模型 SLANeXt，本地已跑通

### CODE REVIEW

**Comment by liu-jiaxuan** [comment]
具体来说，新增了 SLANeXt 的训练能力，相比SLANet_plus获得了较大的性能提高。

通过如下示例，展示SLANeXt 的表格识别能力：
（每组展示中，从上到下为原图、SLANet_plus 识别结果、SLANeXt 识别结果）
1.
image
image
image

2.
image
image
image

3.
6c666092083fd765b60faea69967a10a
84f3f246cd01d30a40fc04b863199536
ac554a66f92ec9ca5e1a09b6f79929d1

File: configs/table/SLANeXt_wired.yml [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  Head:
+    name: SLAHead
+    hidden_size: 512
+    max_text_length: 500
+    loc_reg_num: 8
+
+Loss:
+  name: SLALoss
+  structure_weight: 1.0
+  # SLANeXt默认不进行单元格检测任务的训练, 若需要训练可自行设置loc_weight
```
Comment (cuicheng01) [inline configs/table/SLANeXt_wired.yml:57]: 英文注释

File: configs/table/SLANeXt_wireless.yml [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  Head:
+    name: SLAHead
+    hidden_size: 512
+    max_text_length: 500
+    loc_reg_num: 8
+
+Loss:
+  name: SLALoss
+  structure_weight: 1.0
+  # SLANeXt默认不进行单元格检测任务的训练, 若需要训练可自行设置loc_weight
```
Comment (cuicheng01) [inline configs/table/SLANeXt_wireless.yml:57]: 同上

File: ppocr/modeling/backbones/rec_vary_vit.py [hunk 3]
```diff
@@ -0,0 +1,536 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
```
Comment (cuicheng01) [inline ppocr/modeling/backbones/rec_vary_vit.py:5]: Copyright不对，代码的引用需要补充

File: ppocr/modeling/backbones/rec_vary_vit.py [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+
+import paddle
+import paddle.nn as nn
+import paddle.nn.functional as F
+from typing import Optional, Tuple, Type
+
+from functools import partial
+
+import paddle
+import paddle.nn as nn
```
Comment (cuicheng01) [inline ppocr/modeling/backbones/rec_vary_vit.py:15]: 和上边有重复

File: ppocr/modeling/backbones/rec_vary_vit.py [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+        super().__init__()
+        self.lin1 = nn.Linear(embedding_dim, mlp_dim)
+        self.lin2 = nn.Linear(mlp_dim, embedding_dim)
+        self.act = act()
+
+    def forward(self, x):
+        return self.lin2(self.act(self.lin1(x)))
+
+
+# From https://github.com/facebookresearch/detectron2/blob/main/detectron2/layers/batch_norm.py # noqa
```
Comment (cuicheng01) [inline ppocr/modeling/backbones/rec_vary_vit.py:54]: 如果代码整体是引用自一个地方，这块的引用可以去掉

File: ppocr/modeling/backbones/rec_vary_vit.py [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+
+    def forward(self, x):
+        u = x.mean(1, keepdim=True)
+        s = (x - u).pow(2).mean(1, keepdim=True)
+        x = (x - u) / paddle.sqrt(s + self.epsilon)
+        x = self.weight[:, None, None] * x + self.bias[:, None, None]
+        return x
+
+
+# This class and its supporting functions below lightly adapted from the ViTDet backbone available at: https://github.com/facebookresearch/detectron2/blob/main/detectron2/modeling/backbone/vit.py # noqa
```
Comment (cuicheng01) [inline ppocr/modeling/backbones/rec_vary_vit.py:73]: 同上

File: ppocr/modeling/backbones/rec_vary_vit.py [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+        attn,
+        q,
+        rel_pos_h,
+        rel_pos_w,
+        q_size: Tuple[int, int],
+        k_size: Tuple[int, int],
+):
+    """
+    Calculate decomposed Relative Positional Embeddings from :paper:`mvitv2`.
+    https://github.com/facebookresearch/mvit/blob/19786631e330df9f3622e5402b4a419a263a2c80/mvit/models/attention.py   # noqa B950
```
Comment (cuicheng01) [inline ppocr/modeling/backbones/rec_vary_vit.py:393]: 同上

File: ppocr/modeling/backbones/rec_vary_vit.py [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+        super(Vary_VIT_B, self).__init__()
+
+        self.vision_tower_high = _build_vary(
+                        encoder_embed_dim=768,
+                        encoder_depth=12,
+                        encoder_num_heads=12,
+                        encoder_global_attn_indexes=[2, 5, 8, 11],
+                        image_size=image_size,
+                    )
+        self.out_channels = 1024
```
Comment (cuicheng01) [inline ppocr/modeling/backbones/rec_vary_vit.py:500]: 类内方法之间空一行

File: ppocr/modeling/backbones/rec_vary_vit.py [hunk 9]
```diff
@@ -0,0 +1,536 @@
...
+class Vary_VIT_B(nn.Layer):
+    def __init__(self,  
+        in_channels = 3,
+        image_size=768,
+        encoder_embed_dim=768,
+        encoder_depth=12,
+        encoder_num_heads=12,
+        encoder_global_attn_indexes=[2, 5, 8, 11],
+        ):
+        super(Vary_VIT_B, self).__init__()
```
Comment (cuicheng01) [inline ppocr/modeling/backbones/rec_vary_vit.py:491]: 建议统一使用`super().__init__()`

File: ppocr/modeling/heads/table_att_head.py [hunk 10]
```diff
@@ -268,7 +268,15 @@ def __init__(
         @param max_text_length: max text pred
         """
         super().__init__()
-        in_channels = in_channels[-1]
+
+        if isinstance(in_channels, int):
```
Comment (cuicheng01) [inline ppocr/modeling/heads/table_att_head.py:272]: 这块的逻辑为什么写多行呢？

File: ppocr/losses/table_att_loss.py [hunk 11]
```diff
@@ -22,7 +22,7 @@
 
 
 class TableAttentionLoss(nn.Layer):
-    def __init__(self, structure_weight, loc_weight, **kwargs):
+    def __init__(self, structure_weight=1.0, loc_weight=0.0, **kwargs):
```
Comment (GreatV) [inline ppocr/losses/table_att_loss.py:25]: 这个改了会不会对之前的模型有影响

File: ppocr/losses/table_att_loss.py [hunk 12]
```diff
@@ -58,7 +58,7 @@ def forward(self, predicts, batch):
 
 
 class SLALoss(nn.Layer):
-    def __init__(self, structure_weight, loc_weight, loc_loss="mse", **kwargs):
+    def __init__(self, structure_weight=1.0, loc_weight=0.0, loc_loss="mse", **kwargs):
```
Comment (GreatV) [inline ppocr/losses/table_att_loss.py:61]: 同上

**Review by GreatV — COMMENTED** [review]
https://github.com/PaddlePaddle/PaddleOCR/tree/main/docs/algorithm/table_recognition

最好添加中文文档，介绍新引入的算法模型。

File: ppocr/losses/table_att_loss.py (hunk 12 above)
Comment (liu-jiaxuan) [reply ppocr/losses/table_att_loss.py:61]: 文档后续会进行补充，这几处修改对其他模型无影响

**Review by cuicheng01 — APPROVED** [review]

**Comment by alice77-zhen** [comment]
> 新增最新表格识别模型 SLANeXt，本地已跑通

您好，请问您是怎么跑通的，我本地跑起来的结果是这样显示的：
image

**Comment by alice77-zhen** [comment]
您好，请问有预训练模型吗？

**Comment by alice77-zhen** [comment]
> 新增最新表格识别模型 SLANeXt，本地已跑通

您好，请问得到上面提到的表格结构识别效果大概用了多少训练数据？您有加载什么预训练模型还是从头训的？

**Comment by alice77-zhen** [comment]
导出推理模型的时候报错
错误信息在这里

Traceback (most recent call last):
  File "/home/workspace_disk/PaddleOCR-main/tools/export_model.py", line 37, in <module>
    main()
  File "/home/workspace_disk/PaddleOCR-main/tools/export_model.py", line 33, in main
    export(config)
  File "/home/workspace_disk/PaddleOCR-main/ppocr/utils/export_model.py", line 420, in export
    export_single_model(
  File "/home/workspace_disk/PaddleOCR-main/ppocr/utils/export_model.py", line 306, in export_single_model
    paddle.jit.save(model, save_path)
  File "/usr/local/lib/python3.10/dist-packages/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)
  File "/usr/local/lib/python3.10/dist-packages/paddle/base/wrapped_decorator.py", line 26, in __impl__
    return wrapped_func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/paddle/jit/api.py", line 809, in wrapper
    func(layer, path, input_spec, **configs)
  File "/usr/local/lib/python3.10/dist-packages/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)
  File "/usr/local/lib/python3.10/dist-packages/paddle/base/wrapped_decorator.py", line 26, in __impl__
    return wrapped_func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/paddle/base/dygraph/base.py", line 68, in __impl__
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/paddle/jit/api.py", line 1104, in save
    static_func.concrete_program_specify_input_spec(
  File "/usr/local/lib/python3.10/dist-packages/paddle/jit/dy2static/program_translator.py", line 986, in concrete_program_specify_input_spec
    concrete_program, _ = self.get_concrete_program(
  File "/usr/local/lib/python3.10/dist-packages/paddle/jit/dy2static/program_translator.py", line 875, in get_concrete_program
    concrete_program, partial_program_layer = self._program_cache[
  File "/usr/local/lib/python3.10/dist-packages/paddle/jit/dy2static/program_translator.py", line 1648, in __getitem__
    self._caches[item_id] = self._build_once(item)
  File "/usr/local/lib/python3.10/dist-packages/paddle/jit/dy2static/program_translator.py", line 1575, in _build_once
    concrete_program = ConcreteProgram.from_func_spec(
  File "/usr/local/lib/python3.10/dist-packages/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)
  File "/usr/local/lib/python3.10/dist-packages/paddle/base/wrapped_decorator.py", line 26, in __impl__
    return wrapped_func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/paddle/base/dygraph/base.py", line 68, in __impl__
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/paddle/jit/dy2static/program_translator.py", line 1346, in from_func_spec
    error_data.raise_new_exception()
  File "/usr/local/lib/python3.10/dist-packages/paddle/jit/dy2static/error.py", line 452, in raise_new_exception
    raise new_exception from None
ValueError: In transformed code:

    File "/home/workspace_disk/PaddleOCR-main/ppocr/modeling/architectures/base_model.py", line 84, in forward
        if self.use_backbone:
    File "/home/workspace_disk/PaddleOCR-main/ppocr/modeling/architectures/base_model.py", line 85, in forward
        x = self.backbone(x)
    File "/home/workspace_disk/PaddleOCR-main/ppocr/modeling/backbones/rec_vary_vit.py", line 543, in forward
        cnn_feature = self.vision_tower_high(pixel_values)
    File "/home/workspace_disk/PaddleOCR-main/ppocr/modeling/backbones/rec_vary_vit.py", line 174, in forward
        if self.pos_embed is not None:
    File "/home/workspace_disk/PaddleOCR-main/ppocr/modeling/backbones/rec_vary_vit.py", line 175, in forward
        x = self.patch_embed(x)
        if self.pos_embed is not None:
            x = x + self.pos_embed
            ~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
        for blk in self.blocks:
            x = blk(x)

    File "/usr/local/lib/python3.10/dist-packages/paddle/base/layers/math_op_patch.py", line 595, in __impl__
        current_block(self).append_op(
    File "/usr/local/lib/python3.10/dist-packages/paddle/base/framework.py", line 4467, in append_op
        op = Operator(
    File "/usr/local/lib/python3.10/dist-packages/paddle/base/framework.py", line 3214, in __init__
        self.desc.infer_shape(self.block.desc)

    ValueError: (InvalidArgument) Broadcast dimension mismatch. Operands could not be broadcast together with the shape of X = [-1, 30, 30, 768] and the shape of Y = [1, 32, 32, 768]. Received [30] in X is not equal to [32] in Y at i:1.
  [Hint: Expected x_dims_array[i] == y_dims_array[i] || x_dims_array[i] <= 1 || y_dims_array[i] <= 1 == true, but received x_dims_array[i] == y_dims_array[i] || x_dims_array[i] <= 1 || y_dims_array[i] <= 1:0 != true:1.] (at /paddle/paddle/phi/kernels/funcs/common_shape.h:86)
  [operator < elementwise_add > error]

**Comment by cuicheng01** [comment]
@alice77-zhen 相关权重和推理能力会在paddlex中体现，最近会支持

**Comment by shallow10** [comment]
那个为啥训练的acc的0，一直没有提升

**Comment by Danee-wawawa** [comment]
你好，想问下表格识别模型 SLANeXt相关内容大概什么时候会在paddlex体现哇？

**Pull request merged**